include/bluetooth/services/ots_srv.h

```diff
@@ -0,0 +1,641 @@
+
+	/** Object deleted callback
+	 *
+	 *  This callback is called whenever an object is deleted.
+	 *
+	 *  @param ots  OTS Server instance.
+	 *  @param conn The connection that deleted the object or NULL if
+	 *              this request came from the server.
+	 *  @param id   Object ID.
+	 */
+	void (*obj_deleted)(struct bt_gatt_ots_srv *ots, struct bt_conn *conn,
+			    uint64_t id);
+
+	/** Object selected callback
+	 *
+	 *  This callback is called on successful object selection.
+	 *
+	 *  @param ots  OTS Server instance.
+	 *  @param conn The connection that selected new object.
+	 *  @param id   Object ID.
+	 */
+	void (*obj_selected)(struct bt_gatt_ots_srv *ots, struct bt_conn *conn,
+			     uint64_t id);
+
+	/** Object read callback
+	 *
+	 *  This callback is called multiple times during the Object read
+	 *  operation. OTS module will keep requesting successive Object
+	 *  fragments from the application until the read operation is
+	 *  completed. The end of read operation is indicated by NULL data
+	 *  parameter.
+	 *
+	 *  @param ots    OTS Server instance.
+	 *  @param conn   The connection that read object.
+	 *  @param id     Object ID.
+	 *  @param data   In:  NULL once the read operations is completed.
+	 *                Out: Next chunk of data to be sent.
+	 *  @param len    Remaining length requested by the client.
+	 *  @param offset Object data offset.
+	 *
+	 *  @return Data length to be sent via data parameter. This value
+	 *          should be smaller or equal to the len parameter.
+	 */
+	uint32_t (*obj_read)(struct bt_gatt_ots_srv *ots, struct bt_conn *conn,
+			     uint64_t id, uint8_t **data, uint32_t len,
+			     uint32_t offset);
+};
+
+/** @brief Descriptor for OTS initialization. */
+struct bt_gatt_ots_srv_init {
+	/* OTS features */
+	struct bt_gatt_ots_srv_feat features;
+
+	/* Callbacks */
+	struct bt_gatt_ots_srv_cb *cb;
+};
+
+/** @brief Add an object to the OTS Server.
+ *
+ *  This function adds an object to the OTS database. When the
+ *  object is being added, a callback obj_created() is called
+ *  to notify the user about a new object ID.
+ *
+ *  @param ots      OTS Server instance.
+ *  @param obj_init Meta data of the object.
+ *
+ *  @return 0 in case of success or negative value in case of error.
+ */
+int bt_gatt_ots_srv_object_add(struct bt_gatt_ots_srv *ots,
+			       struct bt_gatt_ots_srv_obj_metadata *obj_init);
+
+/** @brief Delete an object from the OTS Server.
+ *
+ *  This function deletes an object from the OTS database. When the
+ *  object is deleted a callback obj_deleted() is called
+ *  to notify the user about this event. At this point, it is possible
+ *  to free allocated buffer for object data.
+ *
+ *  @param ots OTS Server instance.
+ *  @param id  ID of the object to be deleted (uint48).
+ *
+ *  @return 0 in case of success or negative value in case of error.
+ */
+int bt_gatt_ots_srv_object_delete(struct bt_gatt_ots_srv *ots, uint64_t id);
```

Comment:
I wasn't properly able to determine whether this implementation supports that the application deletes an object while it is being read by a client, and what happens in that scenario. It's a corner case that maybe doesn't need to be handled in this version, but perhaps a TODO should be added if it isn't handled.

Reply:
Its easy to track now since each object has its own state so I added a guard statement that returns error when object is being used (non-idle state).

Reply:
Alright, so the application is not able to delete objects that are being used. Sounds like a good and simple solution